android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/manager/AppServicesStateMachineChecker.kt

```diff
@@ -0,0 +1,105 @@
+     */
+    internal fun handleEvent(event: Event, deviceConfig: DeviceConfig, scopes: Set<String>) {
+        val convertedEvent = when (event) {
+            Event.Account.Start -> FxaEvent.Initialize(
+                ASDeviceConfig(
+                    name = deviceConfig.name,
+                    deviceType = deviceConfig.type.into(),
+                    capabilities = ArrayList(deviceConfig.capabilities.map { it.into() }),
+                ),
+            )
+            is Event.Account.BeginEmailFlow -> FxaEvent.BeginOAuthFlow(ArrayList(scopes), event.entrypoint.entryName)
+            is Event.Account.BeginPairingFlow -> {
+                // pairingUrl should always be non-null, if it is somehow null let's use a
+                // placeholder value that can be identified when checking in sentry
+                val pairingUrl = event.pairingUrl ?: "<null>"
+                FxaEvent.BeginPairingFlow(pairingUrl, ArrayList(scopes), event.entrypoint.entryName)
+            }
+            is Event.Account.AuthenticationError -> FxaEvent.CheckAuthorizationStatus
+            Event.Account.AccessTokenKeyError -> FxaEvent.CheckAuthorizationStatus
+            Event.Account.Logout -> FxaEvent.Disconnect
+            // This is the one ProgressEvent that's considered a "public event" in app-services
+            is Event.Progress.AuthData -> FxaEvent.CompleteOAuthFlow(event.authData.code, event.authData.state)
+            else -> return
+        }
+        rustChecker.handlePublicEvent(convertedEvent)
+    }
+
+    /**
+     * Check a new account state
+     *
+     * Call this after transitioning to an android-components account state.
+     */
+    internal fun checkAccountState(state: AccountState) {
+        val convertedState = when (state) {
+            AccountState.NotAuthenticated -> FxaState.Disconnected
+            is AccountState.Authenticating -> FxaState.Authenticating(state.oAuthUrl)
+            AccountState.Authenticated -> FxaState.Connected
+            AccountState.AuthenticationProblem -> FxaState.AuthIssues
+        }
+        rustChecker.checkPublicState(convertedState)
+    }
+
+    /**
+     * Check an app-services internal state
+     *
+     * The app-services internal states correspond to internal firefox account method calls.  Call
+     * this before making one of those calls.
+     */
+    internal fun checkInternalState(state: FxaStateCheckerState) {
+        rustChecker.checkInternalState(state)
+    }
+
+    /**
+     * Handle an app-services internal event
+     *
+     * The app-services internal states correspond the results of internal firefox account method
+     * calls.  Call this before after making a call.
+     */
+    internal fun handleInternalEvent(event: FxaStateCheckerEvent) {
+        rustChecker.handleInternalEvent(event)
+    }
+}
+
+public val appServicesStateMachineChecker = AppServicesStateMachineChecker()
```

Comment:
it looks like the intention here is to have a singleton. if it's the case, let's just change `AppServicesStateMachineChecker` to an object and use it directly.